File: Week_09-高级动态规划/8.字符串转换整数-atoi.py
#
# @lc app=leetcode.cn id=8 lang=python
#
# [8] 字符串转换整数 (atoi)
#
# 练习写确定有限状态机，详细题解内容见leetcode官方题解
# 类似题目：393. UTF-8 编码验证【中等】
#         65. 有效数字【困难】
# @lc code=start
INT_MAX = 2**31-1
INT_MIN = -2**31
class Automation:
    def __init__(self):
        # 定义初始状态
        self.state = 'start'
        # 初始化符号
        self.sign = 1
        # 初始化数值
        self.ans = 0
        # 定义状态机的状态转换表
        self.table = {
            'start': ['start', 'signed', 'in_number', 'end'],
            'signed': ['end', 'end', 'in_number', 'end'],
            'in_number': ['end', 'end', 'in_number', 'end'],
            'end': ['end', 'end', 'end', 'end'],
        }
    # 获取状态转换过程条件
    def get_col(self, c):
        if c.isspace():
            return 0
        if c == '+' or c == '-':
            return 1
        if c.isdigit():
            return 2
        return 3
    # 根据状态更新数值和符号
    def get(self, c):
        self.state = self.table[self.state][self.get_col(c)]
        if self.state == 'in_number':
            self.ans = self.ans*10 + int(c)
            self.ans = min(self.ans, INT_MAX) if self.sign == 1 else min(self.ans, -INT_MIN)

        if self.state == 'signed':
            self.sign = 1 if c == '+' else -1

class Solution(object):
    def myAtoi(self, s):
        """
        :type s: str
        :rtype: int
        """
        # 初始化自动状态机
        automation = Automation()
        for c in s:
            automation.get(c)
        return automation.sign * automation.ans
# @lc code=end

File: Week_09-高级动态规划/8.字符串转换整数-atoi.cpp
/*
 * @lc app=leetcode.cn id=8 lang=cpp
 *
 * [8] 字符串转换整数 (atoi)-状态机
 */
/*
参考自官方题解，练习使用状态机
*/

// @lc code=start
#include<unordered_map>
#include<vector>
using namespace std;
class Automation {
    string state = "start";
    unordered_map<string, vector<string>> table = {              //用map定义状态机转换规则
        {"start", {"start", "signed", "in_number", "end"}},
        {"signed", {"end", "end", "in_number", "end"}},
        {"in_number", {"end", "end", "in_number", "end"}},
        {"end", {"end", "end", "end", "end"}}
    };
    int get_col(char c) {                       //状态机转换条件
        if (isspace(c)) return 0;               //空字符    字符函数，判断空字符
        if (c == '+' or c == '-') return 1;     //符号字符
        if (isdigit(c)) return 2;               //数组字符  字符函数，判断数字字符
        return 3;                               //其它字符
    }
public:
    int sign = 1;        //符号
    long long ans = 0;   //数值

    //对数字状态和符号状态进行处理
    void get(char c) {
        state = table[state][get_col(c)];
        if (state == "in_number") {             
            ans = ans*10 + (c - '0');
            ans = sign == 1 ? min(ans, (long long)INT_MAX) : min(ans, -(long long)INT_MIN);   //处理溢出
        } else if (state == "signed") {
            sign = c == '+' ? 1 : -1;           //处理符号
        }
    }
};

class Solution {
public:
    int myAtoi(string str) {
        Automation automation;
        for (char c : str) {                //遍历每个字符，触发状态机的状态间转换，以及处理数字状态和符号状态
            automation.get(c);
        }
        return automation.sign * automation.ans;
    }
};
// @lc code=end